### ISSUE TITLE
mumble FreeBSD 10.1 overlay build errors

### ISSUE THREAD

**Issue by feld** [mumble-voip/mumble#1897]
Clang has thrown errors about these. I'm building successfully without the overlay for the moment.

    c++ -Wl,-rpath,/usr/local/lib -fstack-protector -Wl,-z,relro -Wl,-z,now -pthread -Wl,-rpath,/usr/local/lib -shared -o libmanual.so manual.o moc_manual.o  -lQt5Widgets -lQt5Gui -lQt5Core -lGL
    mv -f libmanual.so ../../release/plugins/
    gmake[3]: Leaving directory '/wrkdirs/usr/ports/audio/mumble/work/mumble-82561a4/plugins/manual'
    gmake[2]: Leaving directory '/wrkdirs/usr/ports/audio/mumble/work/mumble-82561a4/plugins'
    cd overlay_gl/ && ( test -e Makefile || /usr/local/lib/qt5/bin/qmake /wrkdirs/usr/ports/audio/mumble/work/mumble-82561a4/overlay_gl/overlay_gl.pro CONFIG+=no-server\ no-alsa\ no-speechd\ no-bundled-speex\ no-bundled-opus\ no-update\ \ bundled-celt\ no-g15\ no-portaudio\ no-pulseaudio\ no-xevie\ no-embed-qt-translations\ no-bonjour -spec /usr/local/lib/qt5/mkspecs/freebsd-clang QMAKE_CC=cc QMAKE_CXX=c++ QMAKE_LINK_C=cc QMAKE_LINK_C_SHLIB=cc QMAKE_LINK=c++ QMAKE_LINK_SHLIB=c++ QMAKE_CFLAGS=-O2\ -pipe\ \ -fstack-protector\ -fno-strict-aliasing QMAKE_CXXFLAGS=-O2\ -pipe\ -fstack-protector\ -fno-strict-aliasing\  QMAKE_LFLAGS=\ -Wl,-rpath,/usr/local/lib\ -fstack-protector QMAKE_LIBS= QMAKE_CFLAGS_DEBUG= QMAKE_CFLAGS_RELEASE= QMAKE_CXXFLAGS_DEBUG= QMAKE_CXXFLAGS_RELEASE= PREFIX=/usr/local CONFIG+=release CONFIG-=debug\ separate_debug_info -o Makefile ) && gmake -f Makefile all
    gmake[2]: Entering directory '/wrkdirs/usr/ports/audio/mumble/work/mumble-82561a4/overlay_gl'
    gmake -f Makefile.Release all
    gmake[3]: Entering directory '/wrkdirs/usr/ports/audio/mumble/work/mumble-82561a4/overlay_gl'
    cc -c -O2 -pipe -fstack-protector -fno-strict-aliasing -fvisibility=hidden -Wall -Wextra -Werror  -Wall -W -fPIC -DRESTRICT=__restrict__ -D_FORTIFY_SOURCE=2 -DTARGET_UNIX -I. -I/usr/local/lib/qt5/mkspecs/freebsd-clang -o release/overlay.o overlay.c
    In file included from overlay.c:694:
    ./init_unix.c:191:3: error: implicit declaration of function 'ElfW' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
                    ElfW(Sym) *symtab = NULL;
                    ^
    ./init_unix.c:191:8: error: use of undeclared identifier 'Sym'
                    ElfW(Sym) *symtab = NULL;
                         ^
    ./init_unix.c:191:14: error: use of undeclared identifier 'symtab'
                    ElfW(Sym) *symtab = NULL;
                               ^
    ./init_unix.c:194:8: error: use of undeclared identifier 'Dyn'; did you mean 'yn'?
                    ElfW(Dyn) *dyn = lm->l_ld;
                         ^~~
                         yn
    /usr/include/math.h:299:8: note: 'yn' declared here
    double  yn(int, double);
            ^
    In file included from overlay.c:694:
    ./init_unix.c:194:14: error: use of undeclared identifier 'dyn'; did you mean 'yn'?
                    ElfW(Dyn) *dyn = lm->l_ld;
                               ^~~
                               yn
    /usr/include/math.h:299:8: note: 'yn' declared here
    double  yn(int, double);
            ^
    In file included from overlay.c:694:
    ./init_unix.c:194:13: error: invalid operands to binary expression ('int' and 'double (*)(int, double)')
                    ElfW(Dyn) *dyn = lm->l_ld;
                    ~~~~~~~~~ ^~~~
    ./init_unix.c:196:10: error: use of undeclared identifier 'dyn'; did you mean 'yn'?
                    while (dyn->d_tag) {
                           ^~~
                           yn
    /usr/include/math.h:299:8: note: 'yn' declared here
    double  yn(int, double);
            ^
    In file included from overlay.c:694:
    ./init_unix.c:196:13: error: member reference base type 'double (int, double)' is not a structure or union
                    while (dyn->d_tag) {
                           ~~~^ ~~~~~
    ./init_unix.c:197:12: error: use of undeclared identifier 'dyn'; did you mean 'yn'?
                            switch (dyn->d_tag) {
                                    ^~~
                                    yn
    /usr/include/math.h:299:8: note: 'yn' declared here
    double  yn(int, double);
            ^
    In file included from overlay.c:694:
    ./init_unix.c:197:15: error: member reference base type 'double (int, double)' is not a structure or union
                            switch (dyn->d_tag) {
                                    ~~~^ ~~~~~
    ./init_unix.c:208:4: error: use of undeclared identifier 'dyn'; did you mean 'yn'?
                            dyn ++;
                            ^~~
                            yn
    /usr/include/math.h:299:8: note: 'yn' declared here
    double  yn(int, double);
            ^
    In file included from overlay.c:694:
    ./init_unix.c:208:8: error: cannot increment value of type 'double (int, double)'
                            dyn ++;
                            ~~~ ^
    ./init_unix.c:210:41: error: use of undeclared identifier 'symtab'; did you mean 'strtab'?
                    ods("Iterating dlsym table %p %p %d", symtab, strtab, nchains);
                                                          ^~~~~~
                                                          strtab
    ./init_unix.c:192:15: note: 'strtab' declared here
                    const char *strtab = NULL;
                                ^
    ./init_unix.c:213:22: error: use of undeclared identifier 'symtab'; did you mean 'strtab'?
                            if (ELF32_ST_TYPE(symtab[i].st_info) != STT_FUNC)
                                              ^~~~~~
                                              strtab
    /usr/include/sys/elf32.h:194:32: note: expanded from macro 'ELF32_ST_TYPE'
    #define ELF32_ST_TYPE(info)             ((info) & 0xf)
                                              ^
    ./init_unix.c:192:15: note: 'strtab' declared here
                    const char *strtab = NULL;
                                ^
    ./init_unix.c:213:31: error: member reference base type 'const char' is not a structure or union
                            if (ELF32_ST_TYPE(symtab[i].st_info) != STT_FUNC)
                                              ~~~~~~~~~^~~~~~~~
    /usr/include/sys/elf32.h:194:32: note: expanded from macro 'ELF32_ST_TYPE'
    #define ELF32_ST_TYPE(info)             ((info) & 0xf)
                                              ^
    In file included from overlay.c:694:
    ./init_unix.c:215:22: error: use of undeclared identifier 'symtab'; did you mean 'strtab'?
                            if (strcmp(strtab+symtab[i].st_name, "dlsym") == 0)
                                              ^~~~~~
                                              strtab
    ./init_unix.c:192:15: note: 'strtab' declared here
                    const char *strtab = NULL;
                                ^
    ./init_unix.c:215:31: error: member reference base type 'const char' is not a structure or union
                            if (strcmp(strtab+symtab[i].st_name, "dlsym") == 0)
                                              ~~~~~~~~~^~~~~~~~
    ./init_unix.c:216:34: error: use of undeclared identifier 'symtab'; did you mean 'strtab'?
                                    odlsym = (void*)lm->l_addr + symtab[i].st_value;
                                                                 ^~~~~~
                                                                 strtab
    ./init_unix.c:192:15: note: 'strtab' declared here
                    const char *strtab = NULL;
                                ^
    ./init_unix.c:216:43: error: member reference base type 'const char' is not a structure or union
                                    odlsym = (void*)lm->l_addr + symtab[i].st_value;
                                                                 ~~~~~~~~~^~~~~~~~~
    19 errors generated.
    Makefile.Release:235: recipe for target 'release/overlay.o' failed
    gmake[3]: *** [release/overlay.o] Error 1
    gmake[3]: Leaving directory '/wrkdirs/usr/ports/audio/mumble/work/mumble-82561a4/overlay_gl'
    Makefile:42: recipe for target 'release-all' failed
    gmake[2]: *** [release-all] Error 2
    gmake[2]: Leaving directory '/wrkdirs/usr/ports/audio/mumble/work/mumble-82561a4/overlay_gl'
    Makefile:268: recipe for target 'sub-overlay_gl-all-ordered' failed
    gmake[1]: *** [sub-overlay_gl-all-ordered] Error 2
    gmake[1]: Leaving directory '/wrkdirs/usr/ports/audio/mumble/work/mumble-82561a4'
    *** Error code 1

**Comment by mkrautz** [comment]
This needs a different implementation on FreeBSD, but the basics should be the same: get a struct link_map for the shared library that contains dlsym(); then look up the address of the original, non-overridden dlsym() symbol.

**Comment by mkrautz** [comment]
Fixed by porting overlay_gl to FreeBSD: 3a356a2a13f39578b9b308bbb9e8970bf9d17978.
